==============================
          BEDTools         
==============================

**Current version**: 2.18.1

Note
-------
Stable release for bedtools were formerly archived on Google Code. Unfortunately, the Google Code
downloads facility is shutting down; so henceforth, all source code and stable releases will be 
maintained via this Github repository.

Summary
-------
Collectively, the bedtools utilities are a swiss-army knife of tools for a wide-range of genomics analysis tasks. The most widely-used tools enable genome arithmetic: that is, set theory on the genome. For example, bedtools allows one to intersect, merge, count, complement, and shuffle genomic intervals from multiple files in widely-used genomic file formats such as BAM, BED, GFF/GTF, VCF.

While each individual tool is designed to do a relatively simple task (e.g., intersect two interval files), quite sophisticated analyses can be conducted by combining multiple bedtools operations on the UNIX command line.

Details
-------
First created through urgency and adrenaline by Aaron Quinlan Spring 2009. 
Maintained by the Quinlan Laboratory at the University of Virginia.

1. **Lead developers**:           Aaron Quinlan, Neil Kindlon
2. **Significant contributions**: Assaf Gordon, Royden Clark, John Marshall, Brent Pedersen, Ryan Dale
3. **Repository**:                https://github.com/arq5x/bedtools2
4. **Stable releases**:           https://github.com/arq5x/bedtools2/releases
5. **Documentation**:             http://bedtools.readthedocs.org
6. **License**:                   Released under GNU public license version 2 (GPL v2).


Citation
--------
*Please cite the following article if you use BEDTools in your research*:
  * Quinlan AR and Hall IM, 2010. BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841â€“842. 

Also, if you use *pybedtools*, please cite the following.
  * Dale RK, Pedersen BS, and Quinlan AR. Pybedtools: a flexible Python library for manipulating genomic datasets and annotations. Bioinformatics (2011). doi:10.1093/bioinformatics/btr539

